zerver/lib/streams.py

```diff
@@ -1152,15 +1167,92 @@ def filter_stream_authorization(
         stream
         for stream in streams
         if stream.id not in {stream.id for stream in unauthorized_streams}
-        and stream.id not in {stream.id for stream in streams_to_which_user_cannot_add_subscribers}
     ]
-    return StreamsCategorizedByPermissions(
+    return BaseStreamsCategorizedByPermissions(
         authorized_streams=authorized_streams,
         unauthorized_streams=unauthorized_streams,
+    )
+
+
+def filter_stream_authorization_for_adding_subscriptions(
+    user_profile: UserProfile, streams: Collection[Stream], is_subscribing_other_users: bool = False
+) -> StreamsCategorizedByPermissionsForAddingSubscribers:
+    if len(streams) == 0:
+        return StreamsCategorizedByPermissionsForAddingSubscribers(
+            authorized_streams=[],
+            unauthorized_streams=[],
+            streams_to_which_user_cannot_add_subscribers=[],
+        )
+
+    user_group_membership_details = UserGroupMembershipDetails(user_recursive_group_ids=None)
+    recipient_ids = [stream.recipient_id for stream in streams]
+    subscribed_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient_id__in=recipient_ids, active=True
+        ).values_list("recipient_id", flat=True)
+    )
+
+    filter_stream_authorization_result = filter_stream_authorization(
+        user_profile, streams, user_group_membership_details, subscribed_recipient_ids
+    )
+
+    streams_to_which_user_cannot_add_subscribers: list[Stream] = []
+    if is_subscribing_other_users:
+        streams_to_which_user_cannot_add_subscribers = (
+            get_streams_to_which_user_cannot_add_subscribers(
+                list(filter_stream_authorization_result.authorized_streams),
+                user_profile,
+                user_group_membership_details=user_group_membership_details,
+            )
+        )
+    authorized_streams = [
+        stream
+        for stream in filter_stream_authorization_result.authorized_streams
+        if stream.id not in {stream.id for stream in streams_to_which_user_cannot_add_subscribers}
+    ]
+    return StreamsCategorizedByPermissionsForAddingSubscribers(
+        authorized_streams=authorized_streams,
+        unauthorized_streams=filter_stream_authorization_result.unauthorized_streams,
         streams_to_which_user_cannot_add_subscribers=streams_to_which_user_cannot_add_subscribers,
     )
 
 
+def filter_stream_authorization_for_content_access(
+    user_profile: UserProfile, streams: Collection[Stream]
+) -> StreamsCategorizedByPermissionsByContentAccess:
```

Comment:
This could've been a function that just returns `content_access_streams` but I thought it would be good to have this function around since this might be a common function to use in future with all 3 stream lists (metadata access, content access and unauthorized)

Reply:
This seems like a really nice return value shape -- nay caller working with `content_access_streams` and `metadata_access_streams` to generate a response will feel fairly readable. 

But I think seeing this, I'd recommend having this implementation replace that of `filter_stream_authorization` rather than making a second function. The fields you have in `StreamsCategorizedByPermissionsByContentAccess` in seem like they would be easily combined into anything a caller of the original `filter_stream_authorization` would want.

Reply:
Not relevant anymore after the last commit changes.